meta/ast_fields.py

```diff
@@ -52,11 +52,18 @@ def main():
             source_file.write(
                 f"{T} ParseTree::{getter}(ParseNode pn) const noexcept {{\n"
                 "    assert(isNode(pn));\n"
+            )
+            if field.property == "selection":
+                source_file.write(f"    assert(({const_ref}).inValidState());\n")
+            source_file.write(
                 f"    return {const_ref};\n"
                 "}\n\n"
             )
             source_file.write(
-                f"void ParseTree::{setter}(ParseNode pn, {T} {field.property}) noexcept {{\n"
+                f"void ParseTree::{setter}(ParseNode pn, {T} {field.property}) noexcept {{\n")
+            if field.property == "selection":
+                source_file.write("    //assert(selection.inValidState()); //DO THIS\n")
```

Comment:
And here you have a reckoning with the design philosophy. Historically, a valid Selection has the left and right parts at the same nesting level. For example, in `x{^2}`, a valid selection would be `[x{^2}]` or `x[{^2}]`, but not `[x{^2]}` or `x{[^2}]`. In the presence of parse errors, it is difficult to ensure that constructed selections will have this form. Likely prohibitively difficult.

Reply:
I've created a bug ticket for this. Since the Typeset::Selection is likely to change rather than the parser, I don't think a fix fits with this MR